java/connectors/semantickernel-connectors-memory-jdbc/src/main/java/com/microsoft/semantickernel/connectors/memory/jdbc/JDBCMemoryStore.java

```diff
@@ -0,0 +1,293 @@
+    public Mono<Void> deleteCollectionAsync(@Nonnull String collectionName) {
+        Objects.requireNonNull(collectionName);
+        return this.dbConnector.deleteCollectionAsync(this.dbConnection, collectionName);
+    }
+
+    @Override
+    public Mono<String> upsertAsync(@Nonnull String collectionName, @Nonnull MemoryRecord record) {
+        Objects.requireNonNull(collectionName);
+        Objects.requireNonNull(record);
+        return doesCollectionExistAsync(collectionName)
+                .handle(
+                        (exists, sink) -> {
+                            if (!exists) {
+                                sink.error(
+                                        new MemoryException(
+                                                ErrorCodes
+                                                        .ATTEMPTED_TO_ACCESS_NONEXISTENT_COLLECTION,
+                                                collectionName));
+                                return;
+                            }
+                            sink.next(exists);
+                        })
+                .then(internalUpsertAsync(collectionName, record));
+    }
+
+    private Mono<String> internalUpsertAsync(
+            @Nonnull String collectionName, @Nonnull MemoryRecord record) {
+        Objects.requireNonNull(collectionName);
+        Objects.requireNonNull(record);
+        try {
+            Mono<Void> update =
+                    this.dbConnector.updateAsync(
+                            this.dbConnection,
+                            collectionName,
+                            record.getMetadata().getId(),
+                            record.getSerializedMetadata(),
+                            record.getSerializedEmbedding(),
+                            record.getTimestamp());
+
+            Mono<Void> insert =
+                    this.dbConnector.insertOrIgnoreAsync(
+                            this.dbConnection,
+                            collectionName,
+                            record.getMetadata().getId(),
+                            record.getSerializedMetadata(),
+                            record.getSerializedEmbedding(),
+                            record.getTimestamp());
+
+            return update.then(insert).then(Mono.just(record.getMetadata().getId()));
+        } catch (JsonProcessingException e) {
+            throw new SQLConnectorException("Error serializing MemoryRecord", e);
+        }
+    }
+
+    @Override
+    public Mono<Collection<String>> upsertBatchAsync(
+            @Nonnull String collectionName, @Nonnull Collection<MemoryRecord> records) {
+        Objects.requireNonNull(collectionName);
+        Objects.requireNonNull(records);
+        return doesCollectionExistAsync(collectionName)
+                .handle(
+                        (exists, sink) -> {
+                            if (!exists) {
+                                sink.error(
+                                        new MemoryException(
```

Comment:
What is a MemoryException?

Reply:
An exception related to any kind of memory issue in the semantic kernel.